### DISCUSSION TITLE
HW1 Part 3

### DISCUSSION THREAD

**Discussion by noah-black1** [cooperrc/computational-mechanics#221] [Q&A]
```python


k_g = 0.013
t = np.linspace(1990,2020,10)
p_num = [np.zeros(len(t))]
p_num[1990] = 1578000000
for i in range(1,len(t)):
    p_num[i]=p_num[i-1]+(k_g*p_num[i-1])
    
plt.plot(t,p_num)
``` 

receiving the following error
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/tmp/ipykernel_201/1415212587.py in <module>
      2 t = np.linspace(1990,2020,10)
      3 p_num = [np.zeros(len(t))]
----> 4 p_num[1990] = 1578000000
      5 for i in range(1,len(t)):
      6     p_num[i]=p_num[i-1]+(k_g*p_num[i-1])

IndexError: list assignment index out of range

**Comment by Ewicko** [comment]
I don't fully understand your index error but I'm using this to access the last/ most recent population number if it helps. (the index [-1] is an easy way to access the last element)


```python 
def euler_integration(t_steps, p0=initial_population, k=k_g):
    '''Euler integration(/numerical?)'''
    populations = [p0]
    for i in range(1, len(t_steps)):
        change_from_step_dp = k * populations[-1] * (t_steps[i] - t_steps[i-1])
        populations.append(populations[-1] + change_from_step_dp)
    return np.array(populations)
```

**Comment by cooperrc** [comment]
> ```python
> k_g = 0.013
> t = np.linspace(1990,2020,10)
> p_num = [np.zeros(len(t))]
> p_num[1990] = 1578000000
> for i in range(1,len(t)):
>     p_num[i]=p_num[i-1]+(k_g*p_num[i-1])
>     
> plt.plot(t,p_num)
> 
> 
> ## receiving the following error
> IndexError Traceback (most recent call last) 
> /tmp/ipykernel_201/1415212587.py in 
> 2 t = np.linspace(1990,2020,10) 
> 3 p_num = [np.zeros(len(t))] 
> ----> 4 p_num[1990] = 1578000000 
> 5 for i in range(1,len(t)): 
> 6 p_num[i]=p_num[i-1]+(k_g*p_num[i-1])
> 
> IndexError: list assignment index out of range
> ```

`IndexError` is coming from the use of `1990` in the indexing position, `p_num[1990]`. It only has 10 spots. You could use `p_num[t == 1990]` for a Boolean index _or_ `p_num[0]` to get `p_num` at 1990
